Move minimal libc malloc to common library

keith-packard opened the pull request:
This will (eventually) permit use of a common malloc implementation with
other C libraries, reducing the amount of Zephyr-specific code required
to support each C library.

This doesn't change the malloc implementation, it just moves the code and renames the
configuration variables.

Closes: #56243

keith-packard committed:
lib/libc: Allow common library to be empty

This allows the C library common library to be empty if no pieces are
needed, eliminating cmake warnings.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
lib/libc: Move malloc from minimal to common library

This will (eventually) permit use of a common malloc implementation with
other C libraries, reducing the amount of Zephyr-specific code required
to support each C library.

Signed-off-by: Keith Packard <keithp@keithp.com>

nordicjm commented on lib/libc/minimal/Kconfig:
```diff
@@ -14,29 +14,30 @@ config MINIMAL_LIBC_NON_REENTRANT_FUNCTIONS
 
 config MINIMAL_LIBC_MALLOC_ARENA_SIZE
-	int "Size of the minimal libc malloc arena"
+	int "Legacy alias for COMMON_LIBC_MALLOC_ARENA_SIZE"
```
This should be deprecated as per https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#deprecated

keith-packard replied:
That doesn't seem to have explicit instructions on how to deprecate a Kconfig variable, but it looks like there are a couple of patterns I can follow in the existing code.

keith-packard replied:
ok, I've added [DEPRECATED] to each affected Kconfig variable, along with references to the new variable.

evgeniy-paltsev commented on lib/libc/minimal/Kconfig:
```diff
@@ -14,29 +14,30 @@ config MINIMAL_LIBC_NON_REENTRANT_FUNCTIONS
 
 config MINIMAL_LIBC_MALLOC_ARENA_SIZE
-	int "Size of the minimal libc malloc arena"
+	int "Legacy alias for COMMON_LIBC_MALLOC_ARENA_SIZE"
```
Prpbably let's just select proper value for `COMMON_LIBC_MALLOC_ARENA_SIZE` based on `MINIMAL_LIBC_MALLOC_ARENA_SIZE`

keith-packard replied:
I stuck this logic into the source code as I couldn't figure out a way to do it in Kconfig -- if `COMMON_LIBC_MALLOC_ARENA_SIZE` is left at the default value of zero, then the code will use the value of `CONFIG_MINIMAL_LIBC_MALLOC_ARENA_SIZE`.

galak reviewed (dismissed):
We should move all the Kconfig's in zephyr from `MINIMAL_LIBC_MALLOC_*` -> to `COMMON_LIBC_MALLOC_*`.

also document deprecation of the `MINIMAL_LIBC_MALLOC_*` Kconfigs in release notes

keith-packard commented:
> We should move all the Kconfig's in zephyr from `MINIMAL_LIBC_MALLOC_*` -> to `COMMON_LIBC_MALLOC_*`.

Thanks, I can do that.

> also document deprecation of the `MINIMAL_LIBC_MALLOC_*` Kconfigs in release notes

Sounds good.

keith-packard committed:
doc/develop: Document common C library APIs

Right now, there's only 'time', which makes this doc pretty short.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
tests/samples: Replace minimal libc malloc configs with common ones

With the minimal C library malloc implementation moving to libc/common, all
of the related Kconfig variables have also changed. Update uses within the
tree.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard commented:
> We should move all the Kconfig's in zephyr from `MINIMAL_LIBC_MALLOC_*` -> to `COMMON_LIBC_MALLOC_*`.

Done.

> also document deprecation of the `MINIMAL_LIBC_MALLOC_*` Kconfigs in release notes

Not sure where these get added?

keith-packard commented:
Lots of updates to clean up this series -- I've got another series stacked on top that switches picolibc -- https://github.com/keith-packard/zephyr/tree/picolibc-common-malloc -- but that required significant logic changes in how the heap is allocated to get alignment and 'use all memory' stuff working. I'll wait to submit that until this series is done, but if anyone is interested in where this is going, that should make it more clear.

keith-packard commented:
It looks like there's a regression caused by a change in the hal/atmel module which makes it incompatible with picolibc. I've submitted a PR to that, https://github.com/zephyrproject-rtos/hal_atmel/pull/32

nordicjm commented on lib/libc/minimal/Kconfig:
```diff
@@ -14,31 +14,43 @@ config MINIMAL_LIBC_NON_REENTRANT_FUNCTIONS
 config MINIMAL_LIBC_CALLOC
-	bool "Minimal libc trivial calloc implementation"
+	bool "[DEPRECATED] Minimal libc trivial calloc implementation"
 	default y
```
To deprecate these, they must not be selected by default, and they must select DEPRECATED

keith-packard replied:
I don't know how to make that compatible with existing uses where this item is enabled by default.

nordicjm replied:
Isn't `COMMON_LIBC_CALLOC` a direct replacement for this, i.e. users would notice no difference by switching? If so, these can be unselected by default and the new ones selected.

keith-packard replied:
But that doesn't seem aligned with the Zephyr deprecation guidelines which ask that existing uses remain supported for two releases -- existing configurations selecting `CONFIG_MINIMAL_LIBC_CALLOC=n` should have calloc disabled, even if they don't include `CONFIG_COMMON_LIBC_CALLOC=n`, right? 

Configurations without any setting for `CONFIG_MINIMAL_LIBC_CALLOC` or `CONFIG_COMMON_LIBC_CALLOC` should have calloc enabled. Suggestions on how to achieve that while `MINIMAL_LIBC_CALLOC` doesn't use `default y` would be most appreciated.

nordicjm replied:
People can still use the existing method, they will just get a deprecation warning when they configure the build, the default configuration does not enable any deprecated options, it would move to the new options. If the existing way is `default y` for the old Kconfig, then use the same for the new or `default y if !MINIMAL_LIBC_MALLOC`, and do not `default y` for the old options

keith-packard replied:
These minimal libc options are a bit unusual -- they would only be specifically configured as 'n', never 'y'. If the default is set to 'n', how do we detect an explicit setting of 'n' so that the existing configuration would continue to work?

nordicjm replied:
The new system should be a drop in replacement for the old, if people want to use the old system, it still needs to be possible, but I would say allowing them to de-select the new options and select the old options is a good enough compromise, that will give them the deprecated option notice, and then after 2 releases the old options can be removed entirely. The new system can be selected by default (I thought the idea was to switch to picolibc as the default anyway?)

keith-packard replied:
I don't know how to meet your goals and still keep existing configurations working. In particular, how would an existing configuration which selects `CONFIG_MINIMAL_LIBC_CALLOC=n` but does not include `CONFIG_COMMON_LIBC_CALLOC=n` work? 

We must be able to distinguish configurations which set `CONFIG_MINIMAL_LIBC_CALLOC=n` from configurations which have no value at all, and I don't know how to do that without using a default value of `y`.

keith-packard replied:
As for moving to picolibc by default, this is part of that effort -- we want to switch picolibc configurations to use the system allocator instead of the picolibc malloc implementation. Having two separate allocators in the system isn't a good use of resources.

nordicjm replied:
@tejlmand for comment/review

nandojve commented:
> It looks like there's a regression caused by a change in the hal/atmel module which makes it incompatible with picolibc. I've submitted a PR to that, [zephyrproject-rtos/hal_atmel#32](https://github.com/zephyrproject-rtos/hal_atmel/pull/32)

Merged!

keith-packard commented:
> Merged!

Thanks! Zephyr PR #56394 queued.

keith-packard commented:
I've temporarily added the atmel patch to this series to see if it fixes the tests. Alas, the wheels appear to have fallen off the CI system today.

evgeniy-paltsev commented:
@keith-packard as polishing this PR may take some time probably it worth to create a separate PR for https://github.com/zephyrproject-rtos/zephyr/issues/56243 fix?

Thanks!

keith-packard commented:
> @keith-packard as polishing this PR may take some time probably it worth to create a separate PR for #56243 fix?
> 
> Thanks!

Feel free; I've got plenty of PRs pending that I don't really want to manage another.

cfriedt reviewed (dismissed):
LGTM

stephanosio commented on lib/libc/common/Kconfig:
```diff
@@ -1,7 +1,13 @@
 # Copyright (c) 2023 Intel Corporation
 # SPDX-License-Identifier: Apache-2.0
 
+config COMMON_LIBC
+	bool
+	help
+	  library of functions shared across libc variants
```
Just to reinforce the existing pattern, can we move `config COMMON_LIBC` to `libc/Kconfig` where all the top-level libc configs exist?

keith-packard replied:
This relates to my comment below; let's resolve that before deciding where this value belongs.

keith-packard replied:
This has been removed instead.

stephanosio commented on lib/libc/common/Kconfig:
```diff
@@ -1,7 +1,13 @@
+	help
+	  library of functions shared across libc variants
+
 config COMMON_LIBC_TIME
 	bool
+	select COMMON_LIBC
 	help
 	  common implementation of time().
```
This should be `depends on COMMON_LIBC` since this is a subordinate configuration of `COMMON_LIBC` (the entirety of this file can be wrapped in `if COMMON_LIBC`).

keith-packard replied:
I'm not sure that's right, or at least doesn't seem to make sense to me. 

We want to let various C library implementations leverage common implementations of Zephyr-related functionality. Each of those will `select` their necessary pieces. Each of those common libc pieces should `select COMMON_LIBC` so that the library itself will be built whenever the selected C library needs something from it -- that's how this fixes #56243.

Note that none of these configuration variables can be selected by the user; they're all purely driven by the C library requirements.

The alternative would be to require that the C library implementations select *both* `COMMON_LIBC` along with the particular bit they need, and that seems redundant as well as error prone; if the C library selects only `COMMON_LIBC`, but not any particular piece, then you raise the issue from #56243. If you select only the piece and not `COMMON_LIBC`, then the library won't get built and the link will fail mysteriously.

stephanosio replied:
Technically, fixing #56243 should only require adding `zephyr_library_property(ALLOW_EMPTY TRUE)`; after all, there is nothing fundamentally wrong with having an empty "library," as long as that is one of the valid configurations.

Now, the question is: what is the purpose of the `COMMON_LIBC` Kconfig and why do we need it?

Since we seem to be making the "common libc" a standalone component (e.g. adding a dedicated doc page for it and exposing its configurations to the users via the common libc-specific Kconfigs like `COMMON_LIBC_MALLOC_ARENA_SIZE`), it would be quite natural to have a top level Kconfig (i.e. `COMMON_LIBC`) that sort of acts like a "master switch," in which case all subordinate Kconfigs should depend on the top level Kconfig (basically, the same idea as `MINIMAL_LIBC`, `NEWLIB_LIBC` and other "top level" libc Kconfigs").

On the other hand, if the `COMMON_LIBC` is only intended to be a helper symbol that indicates one or more common libc services are enabled, then the current implementation selecting `COMMON_LIBC` from the subordinate Kconfigs makes sense; but, this approach will likely require a refactoring in the future -- for instance, what if we decide to add an optional common libc service that is enabled by default (e.g. adding a prompt to `COMMON_LIBC_MALLOC` and making it `default y` such that the common libc malloc is used by default but the libc-specific malloc may also be used as fallback)?

keith-packard replied:
> Technically, fixing #56243 should only require adding `zephyr_library_property(ALLOW_EMPTY TRUE)`; after all, there is nothing fundamentally wrong with having an empty "library," as long as that is one of the valid configurations.
> 
> Now, the question is: what is the purpose of the `COMMON_LIBC` Kconfig and why do we need it?

Only because I was unaware of `zephyr_library_property(ALLOW_EMPTY_TRUE)`. We do not need it if we use that.

> On the other hand, if the `COMMON_LIBC` is only intended to be a helper symbol that indicates one or more common libc services are enabled, then the current implementation selecting `COMMON_LIBC` from the subordinate Kconfigs makes sense; but, this approach will likely require a refactoring in the future -- for instance, what if we decide to add an optional common libc service that is enabled by default (e.g. adding a prompt to `COMMON_LIBC_MALLOC` and making it `default y` such that the common libc malloc is used by default but the libc-specific malloc may also be used as fallback)?

Yup, I'll go remove `COMMON_LIBC` and add `zephyr_library_property(ALLOW_EMPTY_TRUE)` instead.

keith-packard replied:
done. This looks way better. Thanks!

stephanosio commented on doc/develop/languages/c/common_libc.rst:
```diff
@@ -1,4 +1,4 @@
-.. _c_library_common
+.. _c_library_common:
```
This should have been part of the `doc/develop: Document common C library APIs`.

keith-packard replied:
thanks; I missed this in an initial push and amended the wrong patch.

stephanosio commented on doc/develop/languages/c/common_libc.rst:
```diff
@@ -14,3 +14,34 @@ Time function
+:kconfig:option:`CONFIG_COMMON_LIBC_MALLOC` in the application configuration
+file.
+
+The minimal libc internally uses the :ref:`kernel memory heap API <heap_v2>` to
+manage the memory heap used by the standard dynamic memory management interface
+functions such as :c:func:`malloc` and :c:func:`free`.
```
Should mention `common libc` here now.

keith-packard replied:
thanks. sorry for missing these.

stephanosio commented on doc/develop/languages/c/common_libc.rst:
```diff
@@ -14,3 +14,34 @@ Time function
+threads. The size of the internal memory heap is specified by the
+:kconfig:option:`CONFIG_COMMON_LIBC_MALLOC_ARENA_SIZE`.
+
+The default heap size for applications using the minimal C library is zero
+(no heap). For other C library users, if there is an MMU present, then the
+default heap is 16kB. Otherwise, the heap uses all available memory.
```
ditto

stephanosio commented on doc/develop/languages/c/common_libc.rst:
```diff
@@ -14,3 +14,34 @@ Time function
+(:kconfig:option:`COMMON_LIBC_REALLOCARRAY`). Both of these are enabled by
+default as that doesn't impact memory usage in applications not using them.
+
+The standard dynamic memory management interface functions implemented by the
+minimal libc are thread safe and may be simultaneously called by multiple
+threads. These functions are implemented in
+:file:`lib/libc/common/source/stdlib/malloc.c`.
```
ditto

keith-packard replied:
as above

stephanosio commented on doc/develop/languages/c/minimal_libc.rst:
```diff
@@ -26,28 +26,6 @@ C99-compatible formatted output implementation.
 For more details, refer to the :ref:`Formatted Output <formatted_output>` OS
 service documentation.
 
-Dynamic Memory Management
-*************************
-
-Dynamic memory management in the minimal libc can be enabled by selecting the
-:kconfig:option:`CONFIG_MINIMAL_LIBC_MALLOC` in the application configuration
-file.
-
-The minimal libc internally uses the :ref:`kernel memory heap API <heap_v2>` to
-manage the memory heap used by the standard dynamic memory management interface
-functions such as :c:func:`malloc` and :c:func:`free`.
-
-The internal memory heap is normally located in the ``.bss`` section. When
-userspace is enabled, however, it is placed in a dedicated memory partition
-called ``z_malloc_partition``, which can be accessed from the user mode
-threads. The size of the internal memory heap is specified by the
-:kconfig:option:`CONFIG_MINIMAL_LIBC_MALLOC_ARENA_SIZE`.
-
-The standard dynamic memory management interface functions implemented by the
-minimal libc are thread safe and may be simultaneously called by multiple
-threads. These functions are implemented in
-:file:`lib/libc/minimal/source/stdlib/malloc.c`.
```
It would be nice to mention (and link) that the minimal libc uses the "common libc" as the dynamic memory management backend.

keith-packard replied:
A fine plan.

stephanosio commented on lib/libc/common/Kconfig:
```diff
@@ -11,3 +11,35 @@ config COMMON_LIBC_TIME
+
+config COMMON_LIBC_MALLOC
+	bool
+	select COMMON_LIBC
```
same here, should depend on, not select.

stephanosio commented on lib/libc/minimal/Kconfig:
```diff
@@ -14,31 +14,43 @@ config MINIMAL_LIBC_NON_REENTRANT_FUNCTIONS
 	  option may require an additional memory protection region.
 
 config MINIMAL_LIBC_MALLOC
-	bool "Minimal libc malloc implementation"
+	bool "[DEPRECATED] Minimal libc malloc implementation"
 	default y
+	select COMMON_LIBC_MALLOC
 	help
+	  [DEPRECATED] Use COMMON_LIBC_MALLOC
```
Deprecated Kconfigs should `select DEPRECATED`.

keith-packard replied:
I don't know how to make that work. Because this configuration entry (along with the finer grained options below this) is `default y`, applications would probably only ever explicitly set this to `n`. We want to catch those applications and warn them that their configuration selection has been deprecated, as well as warning applications which might explicitly set this to `y` while not warning applications with no setting at all.

stephanosio replied:
Fair enough.

stephanosio commented on samples/bluetooth/peripheral_hr/prj_minimal.conf:
```diff
@@ -80,7 +80,7 @@ CONFIG_ISR_STACK_SIZE=1024
 # Disable features not needed
 CONFIG_TIMESLICING=n
-CONFIG_MINIMAL_LIBC_MALLOC=n
+CONFIG_COMMON_LIBC_MALLOC=n
```
I just realised that we can no longer disable the `malloc` implementation when using the minimal libc after the changes in this PR because `CONFIG_COMMON_LIBC_MALLOC` does not have a prompt (Kconfigs without a prompt is not user configurable and cannot be set here).

keith-packard replied:
odd that we didn't get any errors while testing. Thanks for figuring this out.

stephanosio commented on lib/libc/common/Kconfig:
```diff
@@ -5,3 +5,34 @@ config COMMON_LIBC_TIME
 	help
 	  common implementation of time().
+
+config COMMON_LIBC_MALLOC
+	bool
+	help
+	  common implementation of malloc family
```
_(Continuing from the comment above)_

This needs to have a prompt so that it can be configured in the menuconfig as well as in the conf files (e.g. `CONFIG_COMMON_LIBC_MALLOC=n`.

keith-packard replied:
Yup. All fixed.

stephanosio reviewed (changes requested):
Looks good overall.

Just noticed a minor problem/behaviour difference in  `CONFIG_COMMON_LIBC_MALLOC` that can be easily fixed.

stephanosio commented on lib/libc/minimal/Kconfig:
```diff
@@ -14,31 +14,43 @@ config MINIMAL_LIBC_NON_REENTRANT_FUNCTIONS
 	  option may require an additional memory protection region.
 
 config MINIMAL_LIBC_MALLOC
-	bool "Minimal libc malloc implementation"
+	bool "[DEPRECATED] Minimal libc malloc implementation"
 	default y
+	select COMMON_LIBC_MALLOC
 	help
+	  [DEPRECATED] Use COMMON_LIBC_MALLOC
```
After adding a prompt to `COMMON_LIBC_MALLOC`, this needs to `imply COMMON_LIBC_MALLOC` because, otherwise, you would not be able to do `CONFIG_COMMON_LIBC_MALLOC=n` otherwise.

keith-packard replied:
oh, very fancy. This is the first I've ever heard of the 'imply' keyword. Fixed (and even tested locally).

stephanosio commented:
@keith-packard This needs a rebase.
@galak PTAL (your -1 is still pending).

keith-packard commented:
I rebased, but there weren't any conflicts.

stephanosio commented:
@galak ping

stephanosio commented:
@galak wrote:

> We should move all the Kconfig's in zephyr from `MINIMAL_LIBC_MALLOC_*` -> to `COMMON_LIBC_MALLOC_*`.
> 
> also document deprecation of the `MINIMAL_LIBC_MALLOC_*` Kconfigs in release notes

All `MINIMAL_LIBC_MALLOC_*` configs have been renamed to `COMMON_LIBC_MALLOC_*` with proper deprecation.

The release notes can be updated after the the feature freeze, since more changes related to this will likely follow.
